### PR TITLE
Add heterogeneous model support (per-layer config and modeling)

### PR DESCRIPTION
**Pull request by eladsegal**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=45332) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=45332)

# What does this PR do?

Adds heterogeneous model support - the ability for individual layers to differ from the global config (e.g., different `intermediate_size`, `num_key_value_heads`) and to skip sub-modules entirely (MLP, attention, etc.). This enables models where layers are not uniform, as in pruned, distilled, or NAS-derived architectures.

Examples of such models:

| Model | Derived from |
|---|---|
| [nvidia/Llama-3_3-Nemotron-Super-49B-v1_5](https://huggingface.co/nvidia/Llama-3_3-Nemotron-Super-49B-v1_5) | [meta-llama/Llama-3.3-70B-Instruct](https://huggingface.co/meta-llama/Llama-3.3-70B-Instruct) |
| [nvidia/Llama-3_1-Nemotron-Ultra-253B-v1](https://huggingface.co/nvidia/Llama-3_1-Nemotron-Ultra-253B-v1) | [meta-llama/Llama-3.1-405B-Instruct](https://huggingface.co/meta-llama/Llama-3.1-405B-Instruct) |
| [nvidia/gpt-oss-puzzle-88B](https://huggingface.co/nvidia/gpt-oss-puzzle-88B) | [openai/gpt-oss-120b](https://huggingface.co/openai/gpt-oss-120b/tree/main) |
| [nvidia/NVIDIA-Nemotron-Labs-3-Puzzle-75B-A9B-BF16](https://huggingface.co/nvidia/NVIDIA-Nemotron-Labs-3-Puzzle-75B-A9B-BF16) | [nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-BF16](https://huggingface.co/nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-BF16) |

These models previously required `trust_remote_code=True` to modify the classes of the model they are derived from. With this PR, checkpoints derived from the architectures with built-in support (see [`supported_models.py`](https://github.com/huggingface/transformers/blob/main/src/transformers/integrations/heterogeneity/supported_models.py)) work out of the box, and any other model - including custom models - can opt in with a few lines.

This PR adds the generic modeling integration, skip handling, built-in architecture registry, heterogeneous mask dispatch, and skipped-attention cache handling. It builds on `per_layer_config` introduced in PR #45333, per-layer cache configuration and attention-mask selection from PR #48178, and per-layer MTP configuration from PR #48264.

## How it works

`per_layer_config` describes how individual layers differ from the global configuration. For each layer,
`config.per_layer_config[layer_idx]` provides the resolved configuration, including attribute overrides and `skip`
entries. For example:

```python
from transformers import LlamaConfig

config = LlamaConfig(
    ...,
    per_layer_config={
        0: {"intermediate_size": 64},
        2: {"intermediate_size": 96, "skip": ["attention"]},
    },
)
```

This PR makes supported model layers consume that resolved configuration during construction.

### Modeling (automatic layer patching)

Architecture support is declared with a `HeterogeneousModelingSpec`. Built-in specs live in
[`supported_models.py`](https://github.com/huggingface/transformers/blob/main/src/transformers/integrations/heterogeneity/supported_models.py); any other model opts in by setting `_heterogeneous_modeling_spec` on its `PreTrainedModel` base class.

At model init time, the framework monkey-patches `layer_cls.__init__` to:

1. Resolve the layer index using the spec's `layer_idx_resolver`
2. Pass `config.per_layer_config[layer_idx]` to the original `__init__`
3. For layers with `skip` entries, replace the corresponding sub-modules with parameter-free no-op replacements
4. When `sliding_window` / `attention_chunk_size` vary per layer, patch the layer `forward` to select the attention mask for its layer index

The layer-class wrapper remains installed, but it is active only when a matching model is being initialized. A `ContextVar` scopes the per-model state, which is restored automatically after model initialization by the wrappers installed through `__init_subclass__`.

The spec names the layer class and its layer-index resolver; sub-module skipping additionally requires skip descriptors:

```python
import torch

from transformers.integrations.heterogeneity import (
    HeterogeneousModelingSpec,
    LayerIdxFromArgument,
    ReturnEntry,
    SkipDescriptor,
    get_skip_replacement,
)
from transformers.models.llama.modeling_llama import LlamaAttention, LlamaDecoderLayer, LlamaMLP, LlamaRMSNorm


def identity(hidden_states):
    return hidden_states


spec = HeterogeneousModelingSpec(
    layer_cls=LlamaDecoderLayer,
    layer_idx_resolver=LayerIdxFromArgument("layer_idx"),
    skip_descriptors={
        "attention": SkipDescriptor(
            replacements={
                "input_layernorm": get_skip_replacement(
                    LlamaRMSNorm, ReturnEntry(arg_name="hidden_states", transform=identity)
                ),
                "self_attn": get_skip_replacement(
                    LlamaAttention,
                    [ReturnEntry(arg_name="hidden_states", transform=torch.zeros_like), None],
                ),
            },
            replaces_kv_cache_updater=True,
        ),
        "mlp": SkipDescriptor(
            replacements={
                "post_attention_layernorm": get_skip_replacement(
                    LlamaRMSNorm, ReturnEntry(arg_name="hidden_states", transform=identity)
                ),
                "mlp": get_skip_replacement(LlamaMLP, ReturnEntry(arg_name="x", transform=torch.zeros_like)),
            },
            replaces_kv_cache_updater=False,
        ),
    },
)
```

The skip type names ("attention", "mlp") are what configurations reference in their `skip` lists. Replacement keys are member names, optionally restricted to a member class - `("mixer", NemotronHAttention)` - for layers whose member class varies. `replaces_kv_cache_updater` states whether the skip replaces the member that updates the layer's KV cache (see below).

*Note: a spec without `skip_descriptors` is sufficient for attribute heterogeneity (varying dimensions across layers).*

Generic patching is opt-in. It is applied only when the configuration is heterogeneous and the model declares a `HeterogeneousModelingSpec`. Models without a spec are left unpatched, so architectures that consume `per_layer_config` directly can handle it in their own modeling code.

### MTP models

When the MTP configuration is heterogeneous, this PR adapts the main architecture's skip descriptors to the `mtp_block`, allowing the same generic patching and skips to apply to MTP layers.

### Cache, masks, and KV state

When `sliding_window` or `attention_chunk_size` varies per layer:

- `create_sliding_window_causal_mask` / `create_chunked_causal_mask` return an `AttentionMasksByLayerIdx` mapping keyed by layer index; layers with the same window value share the same generated mask, and each layer's forward selects its own entry
- This layer-indexed mapping composes with `create_masks_for_generate`'s layer-type-keyed output

For layers whose attention is skipped, model initialization records that they do not update the KV cache. `StaticCache` keeps a zero-capacity entry for each such layer so cache indices stay aligned with model layer indices, and ignores those entries when selecting a representative KV layer.

## Key changes

- **`src/transformers/integrations/heterogeneity/` package** - modeling specs and layer-index resolvers, generic layer patching, skip replacements (`SkipDescriptor`, `get_skip_replacement` / `ReturnEntry`), layer-indexed mask construction, built-in architecture registry, configuration validation, and skipped-KV metadata
- **`modeling_utils.py`** - applies generic heterogeneous modeling during model initialization and scopes the initialization context by wrapping subclass `__init__` through `__init_subclass__`
- **`modeling_layers.py`** - adapts the main model's skip descriptors for heterogeneous MTP layers
- **`cache_utils.py`** - gives skipped-attention `StaticCache` layers zero capacity and excludes them from representative-KV-layer selection
- **`masking_utils.py`** - enables heterogeneous dispatch for sliding-window and chunked mask creation
- **`integrations/mxfp4.py`** - fixes homogeneity assumptions in order to support heterogeneous GptOss configurations
- **Docs** - a generic modeling and spec-author guide (`generic_heterogeneous_modeling.md`)

## Tests

Comprehensive test suite in `tests/heterogeneity/` covering 4 architectures (Llama, GptOss, Llama4, NemotronH). It covers structure verification, forward/generate equivalence against manually constructed reference models, model save/load round-trips, heterogeneous cache behavior (including `torch.compile` single-graph decoding), layer-indexed mask construction, MTP models, built-in and custom specs, layer-index resolution, skip replacements, configuration validation, and edge cases.

---
### Concrete example: Llama-3.1-8B-Instruct with 4 layers of attention skipped

<details>
<summary>Code snippet (click to expand)</summary>

```python
from transformers import AutoModelForCausalLM, AutoTokenizer, GenerationConfig, pipeline


base_model_id = "meta-llama/Llama-3.1-8B-Instruct"

# 1. Load model with per-layer overrides (Llama support is built in - no extra setup needed)
model = AutoModelForCausalLM.from_pretrained(
    base_model_id,
    device_map="auto",
    dtype="auto",
    per_layer_config={
        19: {"skip": ["attention"]},
        20: {"skip": ["attention"]},
        22: {"skip": ["attention"]},
        23: {"skip": ["attention"]},
    },
)

# 2. Generate
tokenizer = AutoTokenizer.from_pretrained(base_model_id)
pipe = pipeline(
    "text-generation",
    model=model,
    tokenizer=tokenizer,
)
generation_config = GenerationConfig.from_pretrained(base_model_id)
generation_config.max_new_tokens = 256

messages = [
    {"role": "user", "content": "Hey, how are you doing today?"},
]

outputs = pipe(
    messages,
    generation_config=generation_config,
)

print(outputs[0]["generated_text"][-1]["content"])
```

</details>

## Who can review?
@ArthurZucker
@hmellor

### CODE REVIEW

**Comment by eladsegal** [comment]
@askliar
Related to https://github.com/vllm-project/vllm/pull/36512

**Review by ArthurZucker — COMMENTED** [review]
Hey! thanks for the PR! 
One thing that's super important to stay aligned with `transformers` philo is that is not somthing we want to ship to old models! We'll be super happy to create new models with `modular` that do have heterogenous configs per layer, but adding this to llama break llama. GPT OSS does NOT have this, as such it should not support it. Happy to have `heterogen_gpt_oss` ! 

This removes code paths, and the `if is_heterogeneous` etc. 
Given this, I think it would mostly be a change in the config to have `per_layer_config`. 

We can find a good way to serialize all args in order to reduce the bloat potentially (meaning all attributes are list, at init time this creates a list of config per-layer) this simplifies code changes in `modeling_heterogenous_gpt_oss` for example, that would just get `per_layer_config[i]`

Furthermore I don't understand the `skip` idea can you give a concrete example? you want to skip an entire layer, but you can just set it to `nn.Identity` for example? or just re-index them?

**Comment by hmellor** [comment]
FYI @eladsegal I've already ported Gemma4 in https://github.com/huggingface/transformers/pull/47384 to use the new `per_layer_config` machinery you added. I had to change a couple of small things to make it play nicely with the rest of Transformers, you should check it out!

**Comment by eladsegal** [comment]
> FYI @eladsegal I've already ported Gemma4 in #47384 to use the new `per_layer_config` machinery you added. I had to change a couple of small things to make it play nicely with the rest of Transformers, you should check it out!

Hi @hmellor, that's awesome, thanks for letting me know! Great to see it already has more uses 😃

**Comment by eladsegal** [comment]
Hi @ArthurZucker,

I've made several updates to clean up this PR, keeping heterogeneity-related changes within the heterogeneity package wherever possible.
I've also added documentation, which I hope is clear and helpful.

Looking forward to your comments!

File: src/transformers/integrations/heterogeneity/supported_models.py [hunk 1]
```diff
@@ -22,6 +22,12 @@
 from transformers.integrations.heterogeneity.skip_utils import ReturnEntry, get_skip_replacement
 
 
+MODEL_TYPES_WITH_HETEROGENEOUS_MODELING_PATCHING_DISABLED = {
+    "diffusion_gemma_text",
+    "gemma4_text",
+    "gemma4_unified_text",
+}
+
```
Comment (hmellor) [inline src/transformers/integrations/heterogeneity/supported_models.py]: What does it mean for a model to have patching disabled?

Does it mean "this model is heterogeneous by design, it should be treated that way"?

Comment (eladsegal) [reply src/transformers/integrations/heterogeneity/supported_models.py]: Not quite. It only means that the generic patching mechanism is not supported for this model type. It does not mean that the model supports the full range of heterogeneous overrides that are handled by the generic patching.
For example, these Gemma models explicitly read `head_dim` and `num_key_value_heads` from `config.per_layer_config[layer_idx]`, but still read attributes such as `intermediate_size` from the global config, so it can't vary by layer.

This PR attempts generic patching whenever `per_layer_config` is set on a model's config, but generic patching requires a `HeterogeneousModelingSpec` describing the layer class.
If the model has neither a spec nor an explicit opt-out, an error is raised rather than silently ignoring the per-layer configuration.
I didn't want to force every model using `per_layer_config` to provide a generic spec, so I added the explicit opt-out.

A spec could likely be defined for `gemma4_text` and `gemma4_unified_text`, but support would remain limited: their layer-type lookup requires all resolved configs for layers of the same type to be identical (https://github.com/huggingface/transformers/blob/41db33d8ea9e5aa54f596a6eb1548f651a685c7c/src/transformers/integrations/heterogeneity/configuration_utils.py#L234), which means an override on one of several layers of the same type can fail even when the differing attribute is unrelated to the layer-type-level operation (RoPE in this case).

`diffusion_gemma_text` is more complicated. In addition to that limitation, its custom decoder-mask creation (`create_diffusion_decoder_attention_mask`) may require changes.
More importantly, its encoder and decoder use the same text config but have different layer classes. A spec currently targets only one layer class, so patching one stack leaves the other consuming the same heterogeneous attributes without patching, which can raise `AmbiguousGlobalPerLayerAttributeError`.
In the future, we can add support for specs that need to target multiple layer classes.
